api/src/main/java/org/openmrs/module/fhir/api/util/BaseOpenMRSDataUtil.java

```diff
@@ -0,0 +1,250 @@
+    private static final String VOIDED_BY_URL = "voidedBy";
+    private static final String VOID_REASON_URL = "voidReason";
+    private static final String RETIRED_URL = "retired";
+    private static final String DATE_RETIRED_URL = "dateRetired";
+    private static final String RETIRED_BY_URL = "retiredBy";
+    private static final String RETIRE_REASON_URL = "retireReason";
+
+    private BaseOpenMRSDataUtil() { }
+
+    public static void setBaseExtensionFields(DomainResource fhirResource, BaseOpenmrsData openmrsData) {
+        fhirResource.addExtension(createDateCreatedExtension(openmrsData.getDateCreated()));
+        fhirResource.addExtension(createCreatorExtension(openmrsData.getCreator()));
+
+        if (openmrsData.getDateChanged() != null) {
+            fhirResource.addExtension(createDateChangedExtension(openmrsData.getDateChanged()));
+            fhirResource.addExtension(createChangedByExtension(openmrsData.getChangedBy()));
+        }
+        if (openmrsData.getVoided()) {
+            fhirResource.addExtension(createVoidedExtension(openmrsData.getVoided()));
+            fhirResource.addExtension(createDateVoidedExtension(openmrsData.getDateVoided()));
+            fhirResource.addExtension(createVoidedByExtension(openmrsData.getVoidedBy()));
+            fhirResource.addExtension(createVoidReasonExtension(openmrsData.getVoidReason()));
+        }
+    }
+    
+    public static void setBaseExtensionFields(DomainResource fhirResource, BaseOpenmrsMetadata openmrsMetadata) {
+        fhirResource.addExtension(createDateCreatedExtension(openmrsMetadata.getDateCreated()));
+        fhirResource.addExtension(createCreatorExtension(openmrsMetadata.getCreator()));
+
+        if (openmrsMetadata.getDateChanged() != null) {
+            fhirResource.addExtension(createDateChangedExtension(openmrsMetadata.getDateChanged()));
+            fhirResource.addExtension(createChangedByExtension(openmrsMetadata.getChangedBy()));
+        }
+        if (openmrsMetadata.getRetired()) {
+            fhirResource.addExtension(createRetiredExtension(openmrsMetadata.getRetired()));
+            fhirResource.addExtension(createDateRetiredExtension(openmrsMetadata.getDateRetired()));
+            fhirResource.addExtension(createRetiredByExtension(openmrsMetadata.getRetiredBy()));
+            fhirResource.addExtension(createRetireReasonExtension(openmrsMetadata.getRetireReason()));
+        }
+    }
+
+    public static void readBaseExtensionFields(BaseOpenmrsData openmrsData, DomainResource fhirResource) {
+        for (Extension extension : fhirResource.getExtension()) {
+            setBaseOpenMRSData(openmrsData, extension);
+        }
+    }
+
+    public static void readBaseExtensionFields(BaseOpenmrsMetadata openmrsMetadata, DomainResource fhirResource) {
+        for (Extension extension : fhirResource.getExtension()) {
+            setBaseOpenMRSMetadata(openmrsMetadata, extension);
+        }
+    }
+
+    //region read extensions
```

Comment:
Maybe instead of commenting regions, we could move this logic to a separate class?

Reply:
also, isn't it more ExtensionUtils?

Reply:
Yes, good idea. Those methods will be needed elsewhere in future. Thanks!